### PR TITLE
Smoke test binaries for EL7/glibc-2.17 compatiblity

### PR DESCRIPTION
**Pull request by dekimsey**
## Summary

This adds a quick smoke test of our binaries to verify we haven't exceeded the maximum glibc (2.17) version supported on older OSes (EL7). Which would break our ability to execute on older OSes.

## Why

This is related to our recent efforts to upgrade our GH build environment. When CGO enabled builds are built on Ubuntu-22 they include symbols to glibc versions (2.32+) that exceed our desired minimum OS version (EL7). To mitigate we are instead building on Ubuntu-20. However, Ubuntu-20 itself also includes a newer glibc version. So we need to make sure we check the binaries to prevent us from unintentionally dropping support by adding explicit smoke testing of the binary.

For clarity, here's a breakdown of distribution and glibc versions to provide a bit more context to why we should test on EL7:

| Distro          | glibc version | [EOL](https://endoflife.date)              |
|-----------------|---------------|------------------|
| Debian Jessie   | 2.19          | June 2025        |
| Debian Stretch  | 2.24          | June 2027        |
| CentOS 7        | 2.17          | June 2024        |
| CentOS 8        | 2.28          | December 2021        |
| CentOS 8 Stream       | 2.28          | May 2024        |
| CentOS 9 Stream       | 2.34          | May 2027        |
| RedHat 7        | 2.17          | June 2024        |
| RedHat 8        | 2.28          | May 2031        |
| RedHat 9        | 2.34          | June 2034        |
| Ubuntu 20.04    | 2.31          | May 2030         |
| Ubuntu 22.04    | 2.35          | May 2033         |

## Notes

This only affects CGO_ENABLED=1 binaries (like nomad).

This smoke test is only verifying `amd64` binaries because RH doesn't have an `arm64` container image for EL7 even though it ships the OS (this is fixed as of ubi8). However, no version releases an `arm` image.

## Related

(cross-liking for HC reference)
  - [Findings in hashicorp/nomad-enterprise#951](https://github.com/hashicorp/nomad-enterprise/issues/951#issuecomment-1553338889)
  - [Fixes added to hashicorp/vault-enterprise#4242](https://github.com/hashicorp/vault-enterprise/pull/4242)

### CODE REVIEW

**Review by tgross — APPROVED** [review]
LGTM! Thanks @dekimsey!

Verified in https://github.com/hashicorp/nomad-enterprise/actions/runs/5359393093
and in this PR in https://github.com/hashicorp/nomad/actions/runs/5359921287/jobs/9724253179?pr=17706

I'll merge this in once CI is done.

**Comment by tgross** [comment]
@dekimsey it looks like https://github.com/hashicorp/nomad/pull/17017 got merged before I managed to get this merged and now there's a merge conflict.

**Comment by dekimsey** [comment]
Looks like everything is [clean](https://github.com/hashicorp/nomad/actions/runs/5360236287?pr=17706), so one last push to drop the on.push override used for testing.

image

File: .github/workflows/build.yml
Comment (shoenig) [inline .github/workflows/build.yml:323]: Generally we prefer specific OS's; we've a let a lot of these creep in so I created https://github.com/hashicorp/nomad/issues/17714 to fixup them all.

**Review by shoenig — APPROVED** [review]

File: .github/workflows/build.yml
Comment (dekimsey) [reply .github/workflows/build.yml:323]: Good catch!

**Comment by tgross** [comment]
Looks like this failure https://github.com/hashicorp/nomad/actions/runs/5360327078/jobs/9725065223?pr=17706 was introduced by https://github.com/hashicorp/nomad/pull/17705. And that UI test is a known flake.) Don't worry about that for this PR. Let's get this merged in and I'll fix the other one.

**Comment by tgross** [comment]
Fixed that command test in https://github.com/hashicorp/nomad/pull/17715

**Comment by tgross** [comment]
@gulducat does this PR need to get updated in order to work with the changes we made for permissions/secrets/ENT runners?

**Comment by gulducat** [comment]
@tgross doesn't look to me like it needs anything extra since there are no secrets being used :+1:

**Pull request merged**
